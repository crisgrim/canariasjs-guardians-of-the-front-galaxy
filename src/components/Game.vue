<template>
  <li
    class="game"
    @click="handleClick"
  >
    <img :src="game.image">

    <div class="content">
      <h2>{{ game.title }}, {{ game.year }}</h2>
      <h3>Published: {{ formatPublishDate }}</h3>

      <p>
        Started:
        <i :class="[game.started ? 'fas fa-check' : 'fas fa-times']" />
        Ended:
        <i :class="[game.ended ? 'fas fa-check' : 'fas fa-times']" />
      </p>

      <p>Percentage {{ game.percentage }}</p>

      <p>Estimated {{ game.estimated }}</p>
      <p>Score {{ game.score }} - Hype {{ game.hype }}</p>

      <div class="categories">
        <p
          v-for="categorie in game.categories"
          :key="categorie"
        >
          {{ categorie }}
        </p>
      </div>
    </div>
  </li>
</template>

<script>
import formatDate from '@/utils/format-date';

export default {
  name: 'Game',
  props: {
    game: {
      type: Object,
      required: true,
    },
  },
  computed: {
    formatPublishDate() {
      return formatDate(this.game.publish);
    },
  },
  methods: {
    handleClick() {
      this.$emit('game-clicked', this.game);
    },
    icon(isDone) {
      const icon = isDone ? 'check' : 'times';
      return `<i class="fas fa-${icon}"></i>`;
    },
  },
};
</script>

<style scoped lang="scss">
li {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  margin-bottom: 20px;
}

.categories {
  display: flex;
  flex-direction: row;
  font-size: 13px;
  color: #ffffff;
}

.categories p {
    padding: 0px 15px;
    margin: 2px;
    background-color: black;
}

h2 {
  margin: 0 0 5px;
}

h3 {
  margin-top: 0;
}

.content {
  padding-left: 40px;
}
</style>
