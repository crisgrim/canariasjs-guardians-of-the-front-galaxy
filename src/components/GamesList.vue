<template>
  <ul>
    <Game
      v-for="game in games"
      :key="game.id"
      :game="game"
      @game-clicked="handleGameClick"
    />
  </ul>
</template>

<script>
import Game from '@/components/Game.vue';

export default {
  name: 'GameList',
  components: {
    Game,
  },
  props: {
    games: {
      type: Array,
      required: true,
    },
  },
  methods: {
    handleGameClick() {
      // Do something
    },
    /*
    handleGameClick(game) {
      console.log(game);
    },
    */
  },
};
</script>

<style lang="scss">
ul {
  margin: 0;
  padding: 0;
  text-align: left;
}
</style>
