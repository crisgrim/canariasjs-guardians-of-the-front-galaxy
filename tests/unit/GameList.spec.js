import { mount } from '@vue/test-utils';
import GamesList from '../../src/components/GamesList';
import Game from '../../src/components/Game';

describe('GamesList.spec.js', () => {
  const games = [{
    id: 0,
    title: 'God of war',
    year: 2018,
    publish: new Date(2018, 11, 1),
    started: false,
    ended: false,
    percentage: 0,
    estimated: 20,
    score: 8,
    hype: 10,
    categories: ['Adventure', 'Vikings', 'One Player'],
    image: 'https://static.promodescuentos.com/pepperpdimages/threads/thread_large/default/251090_1.jpg',
  }];

  let cmp;

  beforeEach(() => {
    cmp = mount(GamesList, {
      propsData: {
        games,
      },
    });
  });

  it('has received games as property', () => {
    expect(cmp.props().games).toEqual(games);
  });

  it('has the expected html structure', () => {
    expect(cmp.element).toMatchSnapshot();
  });

  it('is a GamesList component', () => {
    expect(cmp.is(GamesList)).toBe(true);
  });

  it('contains a game component', () => {
    expect(cmp.contains(Game)).toBe(true);
  });

  // Vue instance
  it('Both GamesList and game are vue instances', () => {
    expect(cmp.isVueInstance()).toBe(true);
    expect(cmp.find(Game).isVueInstance()).toBe(true);
  });

  // Structure
  it('game element exists', () => {
    expect(cmp.find('.game').exists()).toBe(true);
  });

  it('game is not empty', () => {
    expect(cmp.find(Game).isEmpty()).toBe(false);
  });

  it('game has a class attribute set to "game"', () => {
    expect(cmp.find(Game).attributes().class).toBe('game');
  });

  it('Calls handleGameClick when @game-click happens', () => {
    const stub = jest.fn();
    cmp.setMethods({ handleGameClick: stub });
    cmp.find('.game').vm.$emit('game-clicked', {
      id: 0,
      title: 'God of war',
      year: 2018,
      publish: new Date(2018, 11, 1),
      started: false,
      ended: false,
      percentage: 0,
      estimated: 20,
      score: 8,
      hype: 10,
      categories: ['Adventure', 'Vikings', 'One Player'],
      image: 'https://static.promodescuentos.com/pepperpdimages/threads/thread_large/default/251090_1.jpg',
    });

    expect(stub).toBeCalledWith({
      id: 0,
      title: 'God of war',
      year: 2018,
      publish: new Date(2018, 11, 1),
      started: false,
      ended: false,
      percentage: 0,
      estimated: 20,
      score: 8,
      hype: 10,
      categories: ['Adventure', 'Vikings', 'One Player'],
      image: 'https://static.promodescuentos.com/pepperpdimages/threads/thread_large/default/251090_1.jpg',
    });
  });
});
