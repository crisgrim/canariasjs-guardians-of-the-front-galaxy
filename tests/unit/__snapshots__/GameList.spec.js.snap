// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamesList.spec.js has the expected html structure 1`] = `
<ul>
  <li
    class="game"
  >
    <img
      src="https://static.promodescuentos.com/pepperpdimages/threads/thread_large/default/251090_1.jpg"
    />
     
    <div
      class="content"
    >
      <h2>
        God of war, 2018
      </h2>
       
      <h3>
        Published: 1/11/2018
      </h3>
       
      <p>
        
      Started:
      
        <i
          class="fas fa-times"
        />
        
      Ended:
      
        <i
          class="fas fa-times"
        />
      </p>
       
      <p>
        Percentage 0
      </p>
       
      <p>
        Estimated 20
      </p>
       
      <p>
        Score 8 - Hype 10
      </p>
       
      <div
        class="categories"
      >
        <p>
          
        Adventure
      
        </p>
        <p>
          
        Vikings
      
        </p>
        <p>
          
        One Player
      
        </p>
      </div>
    </div>
  </li>
</ul>
`;
